### DISCUSSION TITLE
Is there a way to wait for the response to a request?

### DISCUSSION THREAD

**Discussion by stephe-ada-guru** [joaotavora/eglot#1088] [Q&A]
I'm experimenting with using SemanticTokens for font-lock, using a fork of https://github.com/AkibAzmain/eglot/tree/semantic-tokens. I have test code that checks for the result, by looking at the faces of various identifiers in a file.

The test calls font-lock-ensure, which ensures that the SemanticTokens request for the region containing the identifier test is sent, then checks the face of the identifier. The test is failing, because it does not wait for the server response.

Is there a way to get the last request id, and wait for the corresponding response? This is only for a test, so it can be somewhat hacky, but I'd like something better than "wait 0.1 seconds".

**Comment by nemethf** [comment]
Can you redefine `jsonrpc-async-request` with `jsonrpc-request`?  Or with something that calls `jsonrpc-request`?

**Reply by joaotavora** [reply]
I'm not sure that's a great idea.  @stephe-ada-guru I'd suggest to look at the macros in `eglot-tests.el`.  There is a moderately useful `eglot--wait-for` there macro that allows you to sniff the LSP traffic and wait for a specific condition, with suitable error reporting (hopefully).

**Reply by stephe-ada-guru** [reply]
That might work, especially if I check the request, and only do jsonrpc-request for  :textDocument/semanticTokens/range. Hmm; jsonrpc-async-request takes a :success-fn arg, so it's probably simpler to redefine eglot--semantic-tokens-highlight-range.
